rfcs/20231129-tuning-api/README.md

```diff
@@ -0,0 +1,134 @@
+# RFC for Auto-Tuning API
+
+## Introduction & Motivation
+
+Auto-tuning is a feature supported as a mode in PyTorch's [torch.compile](https://pytorch.org/docs/stable/generated/torch.compile.html) function 
+and in TensorFlow via [XLA](https://github.com/sourcecode369/tensorflow-1/blob/9f446aba8aaeb2b3c4c6e5ba1ab4cf31494b8a64/tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc#L279).
+While the median model level improvement is generally modest, there are cases where there are 
+large speedups. This has been observed on Intel hardware (see [Using tuning to evaluate the level of oneDNN performance](https://github.com/intel-innersource/libraries.performance.math.onednn/pull/5931)) 
+and externally on other [hardware](https://mangpo.net/papers/xla-autotuning-pact2021.pdf).
+
+The goal is to allow end users to optionally try auto-tuning in the case where out of the
+box performance is insufficient.
+
+## Proposal
+OneDNN should implement auto tuning as a feature that can be exposed to end users of frameworks.
+Major requirements for framework integration:
+1) No changes to primitive API.
+2) Primitive and kernel cache states should not be affected.
+3) No regressions after tuning.
+4) Simple knob to enable/disable tuning.
+
+### Option 1- (Recommended)
+Basic Usage Scenario
+```c
+create_primitive(prim); 
+dnnl_set_tune(true);   
+execute(prim);  //tuning happens here
```

Comment:
This scenario looks like the initially created primitive goes unused (as it's not needed for following tuning), and we have unnecessary overhead. Is it correct? Is there a way to avoid that?

Reply:
It can be avoided if the tuning state is set before primitive creation or the proposed `dispatch_mode_t` is used.
On the other hand, there will be many primitive creations during the tuning process itself, so i don't anticipate the overhead from the first primitive creation to be a big overhead.

Reply:
If there is no reason to keep the flow in this order (create -> set_tune) then I think it makes sense to flip these steps via an attribute. In general the API would work like this:

- Add a tuning knob to the primitive attribute to specify that primitive is created for tuning (let's call this primitive a "tunable" primitive)
- If function API is explicitly requested - introduce `dnnl_set_tune()` which will modify the attribute on creation automatically
- During execution of a tunable primitive (`execute()` call) oneDNN performs all necessary benchmarking or no benchmarking (e.g. when the primitive doesn't support tuning or when this case was already tuned). oneDNN also doesn't guarantee correctness upon execution for tunable primitives
- A tunable primitive can't be used for normal execution, it must be re-created by the user (like in your example)

Do you see any issues with such API?

Reply:
I thought about following the same approach as for FP math mode:

https://github.com/oneapi-src/oneDNN/blob/c900643a3f87908679d6780476af1447ea5cb35b/src/common/primitive_attr.hpp#L620

We have a global setter which sets a global value that is used as the default value during primitive attribute creation.

Reply:
I see. Does this imply after tuning is unset, the `primitive_attr_t`  needs to be recreated?